cmd/juju/cloud/updatecredential.go

```diff
@@ -101,57 +136,244 @@ func (c *updateCredentialCommand) getAPI() (credentialAPI, error) {
-		return err
+	if c.Local {
+		return c.updateLocalCredentials(ctx, credentials)
+	}
+	return c.updateRemoteCredentials(ctx, credentials)
+}
+
+func credentialsFromFile(credentialsFile, cloudName, credentialName string) (map[string]jujucloud.CloudCredential, error) {
+	data, err := ioutil.ReadFile(credentialsFile)
+	if err != nil {
+		return nil, errors.Annotate(err, "reading credentials file")
+	}
+	specifiedCredentials, err := jujucloud.ParseCredentials(data)
+	if err != nil {
+		return nil, errors.Annotate(err, "parsing credentials file")
+	}
+
+	if cloudName == "" && credentialName == "" {
+		return specifiedCredentials, nil
+	}
+
+	filteredByCloud := map[string]jujucloud.CloudCredential{}
+	if cloudName != "" {
+		cloudCredentials, ok := specifiedCredentials[cloudName]
+		if !ok {
+			return nil, errors.Errorf("No credentials for cloud %q exist in file %q.", cloudName, credentialsFile)
+		}
+		filteredByCloud[cloudName] = cloudCredentials
+	} else {
+		filteredByCloud = specifiedCredentials
+	}
+	if credentialName == "" {
+		return filteredByCloud, nil
+	}
+
+	filteredByName := map[string]jujucloud.CloudCredential{}
+	for aCloud, cloudCredentials := range filteredByCloud {
+		for name, aCredential := range cloudCredentials.AuthCredentials {
+			if name == credentialName {
+				filteredByName[aCloud] = jujucloud.CloudCredential{
+					AuthCredentials: map[string]jujucloud.Credential{name: aCredential},
+					DefaultRegion:   cloudCredentials.DefaultRegion,
+				}
+			}
+		}
+	}
+
+	if len(filteredByName) == 0 {
+		return nil, errors.Errorf("No credential %q for cloud %q exists in file %s.", credentialName, cloudName, credentialsFile)
 	}
+	return filteredByName, nil
+}
+
+func credentialsFromLocalCache(store jujuclient.ClientStore, cloudName, credentialName string) (map[string]jujucloud.CloudCredential, error) {
+	cloudCredentials, err := store.CredentialForCloud(cloudName)
+	if err != nil {
+		return nil, errors.Annotate(err, "loading credentials")
+	}
+
+	found := map[string]jujucloud.CloudCredential{}
+	if credentialName == "" {
+		found[cloudName] = *cloudCredentials
+		return found, nil
+	}
+	for name, aCredential := range cloudCredentials.AuthCredentials {
+		if name == credentialName {
+			found[cloudName] = jujucloud.CloudCredential{
+				AuthCredentials: map[string]jujucloud.Credential{name: aCredential},
+				DefaultRegion:   cloudCredentials.DefaultRegion,
+			}
+			return found, nil
+		}
+	}
+	return nil, errors.Errorf("No credential %q for cloud %q exists in local client cache.", credentialName, cloudName)
+}
+
+func (c *updateCredentialCommand) updateLocalCredentials(ctx *cmd.Context, update map[string]jujucloud.CloudCredential) error {
+	erred := false
+	for cloudName, cloudCredentials := range update {
+		aCloud, err := common.CloudByName(cloudName)
+		if errors.IsNotFound(err) {
+			ctx.Infof("Cloud %q not found.", cloudName)
+			erred = true
+			continue
+		} else if err != nil {
+			logger.Errorf("%v", err)
```

Comment:
Is there a reason not to use cmd.WriteError() here?

Reply:
No reason besides consistency... There are 5 references to cmd.WriteError() in th whole of cmd/juju and 100+ to logger.